%!TEX root = ../thesis.tex

\section{Introduction}\label{section:introduction}

Reconstructing character-based phylogenies, which are used to study the evolution of characters shared by a collection of species (taxa or individuals), is a recurring problem in Bioinformatics.
However, these problems are still not easily solvable \cite{DPQ2011}.
In fact, there exist multiple approaches to evolutionary history reconstruction \cite{SD2003, MPL2004, BI2001}; most of those approaches try to reduce the problem by limiting the way characters can change state in the phylogenetic tree \cite{CS1990, PPptime1994}.

Each problem requires proper approaches and models.
Parsimony methods gained relevance in recent years, with their new applications being haplotyping \cite{PPHltime2005, PPHltime2007}, protein domains/networks analysis and genetic markers analysis.
One of the most widely studied character-based parsimony models is the Perfect Phylogeny model, which assumes that no character can mutate more than once in the phylogenetic tree.

Recent applications like carcinogenesis, i.e., the process by which normal cells are transformed into cancer cells, can find only a partial solution in perfect phylogenies \cite{PPE2014}.
More general models are required to represent the possibility of back mutations.
However, generalized models such as the Dollo and the Camin-Sokal lead to NP-Hard computational problems.

The Persistent Phylogeny model \cite{PPPbin2012}, despite being a generalization of the Perfect Phylogeny model, preserves its computational tractability; in fact, a polynomial time algorithm has recently been introduced \cite{PPPptime2016}.
This model allows characters to be acquired and lost at most once in the evolutionary history, returning back to the ancestral state.
Characters that are acquired and lost in a persistent phylogeny are called persistent, and they represent a special case of back mutations, which has been considered in the study of the evolution of protein domains \cite{CPPP2014}.

\subsection{Character evolution in Bioinformatics}\label{section:character-evolution}

To study the mutations of characters in an evolutionary tree we can represent species and their characters as rows and columns of a matrix; each element of the matrix marks the specific state of a character for a species.

We will consider the case in which all characters are binary, making them able to assume only one of two states, 0 or 1.
For each species, the state of a character represents if the species has or doesn't have a given feature.

Binary phylogeny models have already been explored \cite{PPPbin2012, NPP2007, IDPP2004} and used to produce algorithms reaching linear-time computational complexity \cite{PPHltime2007, PPHltime2005}.

\begin{definition}\label{definition:m}
  \m{} is a $n \times m$ binary matrix over a set $m$ of characters and a set $n$ of species.

  \m[i][j] represents the state of a character \character[j] for a species \species[i].
\end{definition}

A matrix \m{} can then be represented as an undirected bipartite graph.
The set of vertices of the graph is $V = S \cup C$, where $S$ is the set of species and $C$ the set of characters.

\begin{figure}[ht]
  \input{figures/figure1}

  \caption{An instance of a character-based phylogeny represented as matrix and graph $(\text{4 species} \times \text{4 characters})$}\label{figure:1}
\end{figure}

In the figure above (\ref{figure:1}) we have a set of species $S = \{ \species[0], \species[1], \species[2], \species[3] \}$ and a set of characters $C = \{ \character[0], \character[1], \character[2], \character[3] \}$.
Each element $\m[i][j] = 1$ of the matrix (\ref{figure:1:a}) is shown as the edge \edge{\species[i]}{\character[j]} in the corresponding graph (\ref{figure:1:b}).

\subsubsection{Character states}\label{section:character-states}

Mutations in a phylogenetic tree can be represented as \emph{signed} characters, using the notation \character[][+] for acquired characters and \character[][-] for lost characters.

Since the Persistent Perfect Phylogeny model (PPP for short) enables characters to be lost at most once in the evolutionary events, we keep track of the ones that were previously acquired introducing the concept of \emph{active} characters \cite{PPPbin2012, PPPcgraph2016, PPPptime2016}.

Active characters are a subset of the instance's current set of characters $C = \{ \character[0], \dots, \character[m-1] \}$.
A character becomes active whenever it is gained in the tree.

\subsection{Persistent Phylogeny problem}\label{section:ppp}

Each instance of a PPP is associated to a pair \ma{} and a rooted tree $T$, where $T$ is a \emph{persistent phylogeny} for the pair \ma{}.

We now formalize input and output parameters for the Persistent Phylogeny problem \cite{PPPptime2016}.

\begin{definition}[Persistent Phylogeny problem]\label{definition:ppp}
  \text{}

  \textit{Input:} pair \ma{} where \m{} is a $n \times m$ binary matrix over a set of $m$ characters and a set of $n$ species, and A is a subset of its characters.

  \textit{Output:} persistent phylogeny tree $T$ solving \m{} if it exists.
\end{definition}

If \ma{} admits a persistent phylogeny we then say that the matrix $M$ is solved by the tree $T$ \cite{PPPptime2016,PPPcgraph2016}.

\subsubsection{Red-black graph representation}\label{section:grb}

A red-black graph for an instance of PPP is an undirected bipartite graph whose edges are colored as either red or black.

\begin{definition}[Red-black graph for PPP]\label{definition:grb}
  Let $S$ be a set of species vertices, $C$ a set of character vertices, $B$ a set of black edges and $R$ a set of red edges.
  Then \grb{} is defined as follows:

  \[ \grb{} = (S \cup C, B \cup R) \]

  The vertex set $V$ of \grb{} is formally represented as two disjoint and independent sets $S$ and $C$.

  The edge set $E$ of \grb{} is formally represented as two disjoint and independent sets $B$ and $R$.
\end{definition}

The set of characters adjacent to a species \species{} is denoted as $C(\species{})$.
Conversely, we use $S(\character{})$ to indicate the set of species adjacent to a character \character{}.

A character \character{} is \emph{inactive} in \grb{} if all of its incident edges are black.
Conversely, \character{} is \emph{active} in \grb{} if all of its incident edges are red.

A character \character{} is \emph{universal} in \grb{} if it's inactive and its set $S(\character{})$ is made up of all the species in the connected component that contains \character{}.
Moreover, \character{} is \emph{free} in \grb{} if it's active and $S(\character{})$ is made up of all the species in the connected component that contains \character{}.

We say that a red-black graph \grb{} is solved by a tree $T$ if $T$ is a persisteny phylogeny for the graph.
Moreover, the construction of a tree $T$ can be expressed as a series of graph operations performed on the red-black graph.
This sequence of operations is called \emph{c-reduction}; a c-reduction is then called \emph{successful c-reduction} if, when applied to a red-black graph that admits a persistent phylogeny, results in an empty graph.

Each graph operation in a c-reduction, represented as a signed character, is called a \emph{realization}.

\begin{definition}[Realization]\label{definition:realization}
  Let \grb{} be a red-black graph and \character{} a character of \grb{}.
  Let $Conn(\character{})$ be the set of species of \grb{} in the connected component that contains \character{}.

  The realization of \character[][+] on \grb{}, which is feasible only if \character{} is inactive, involves adding red edges between \character{} and each species in $Conn(\character{}) \setminus S(\character{})$ and deleting all black edges incident on \character{}.

  The realization of \character[][-] on \grb{}, which is feasible only if \character{} is active and there is no species in $Conn(\character{}) \setminus S(\character{})$ (i.e., \character{} is free), involves deleting all (red) edges incident on \character{}.

  After a realization is performed on \grb{}, all isolated vertices of \grb{} may be deleted.
\end{definition}

The realization of a species \species{} is defined as the realization of each character in $C(\species{})$ in any order.

The notion of c-reduction was first introduced in \cite{PPPbin2012} and is defined as a sequence of positive characters \character[][+].
A c-reduction $R$ is \emph{feasible} for \grb{} if the realization of each signed character in $R$ is feasible for the current \grb{}, and the realization of \character[i][-] is applied whenever a character \character[i] becomes free \cite{PPPcgraph2016}.

A c-reduction $\langle \character[0][\pm], \dots, \character[k][\pm] \rangle$ (where $0$ and $k$ represent the indexes of the signed characters in the sequence) applied to \grb{} produces a graph \grb[k], which represents the state of \grb{} after the realization of the characters from \character[0][\pm] to \character[k][\pm].
The realization of \character[0][\pm] on \grb{} produces \grb[0], while for indexes $i > 0$ each \grb[i] is generated from the realization of \character[i][\pm] applied on \grb[i-1].

We provide an example for a c-reduction being applied to a red-black graph in figure \ref{figure:3}.

\subsubsection*{Red \boldmath{\sg{}}s}

For a c-reduction to be successful, the graphs \grb[i] induced by the realization of each signed character \character[i][\pm] must not contain a red \sg{}.

We provide some graphical examples of red \sg{}s in figure \ref{figure:2}, where we can observe (\ref{figure:2:a} and \ref{figure:2:c} in particular) that none of the characters are free in \grb{}, making the realization of any \character[i][-] impossible; then, the graph can't be emptied by a c-reduction.

\begin{figure}[hp]
  \input{figures/figure2}

  \caption{Examples of red \sg{}s}\label{figure:2}
\end{figure}

\begin{definition}[Red \sg{}]\label{definition:sigma-graph}
  Let \grb{} be a red-black graph, \character[0] \character[1] two active characters of \grb{} and \species[0] \species[1] \species[2] three species of \grb{}.

  \grb{} contains a red \sg{} if there exists a (red) path $\species[0], \character[0], \species[1], \character[1], \species[2]$ of length 4, and the edges \edge{\character[0]}{\species[2]} and \edge{\character[1]}{\species[0]} do not exist.
\end{definition}

A c-reduction $R$ is successful if, when applied to a red-black graph \grb{}, the graphs \grb[i] induced by the realization of each signed character \character[i][\pm] of $R$ does not contain a red \sg{}.

A red-black graph consisting of \textit{k} connected components has a successful reduction $R$ if and only if each component has a successful reduction $R_{i}$ (with $1 < i < k$).
Then $R$ consists of any concatenation of the \textit{k} sequences $R_{i}$ \cite{PPPptime2016}.

\begin{figure}[hp]
  \input{figures/figure3}

  \caption{Application of a c-reduction $R = \langle \character[3][+], \character[1][+], \character[2][+], \character[3][-], \character[0][+], \character[1][-], \character[2][-] \rangle$ to a red-black graph \grb{}}\label{figure:3}
\end{figure}
